timemachine/fe/atom_mapping.py

```diff
@@ -7,64 +8,64 @@
-    """
-
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-
-    def compare(self, p, mol1, atom1, mol2, atom2):
+def mcs(a, b, threshold: float = 2.0, timeout: int = 5, smarts: Optional[str] = None, conformer_aware=True, retry=True):
+    """Find maximum common substructure between mols a and b
+    using reasonable settings for single topology:
+    * disallow partial ring matches
+    * disregard element identity and valence
 
-        if mol1.GetAtomWithIdx(atom1).GetDegree() == 1:
-            return False
-        if mol2.GetAtomWithIdx(atom2).GetDegree() == 1:
-            return False
+    if conformer_aware=True, only match atoms within distance threshold
+        (assumes conformers are aligned)
 
-        x_i = mol1.GetConformer(0).GetPositions()[atom1]
-        x_j = mol2.GetConformer(0).GetPositions()[atom2]
-
-        threshold = 1.0  # angstroms
-        return bool(np.linalg.norm(x_i - x_j) <= threshold)
-
-
-def mcs_map(a, b, threshold: float = 2.0, timeout: int = 5, smarts: Optional[str] = None):
-    """Find the MCS map of going from A to B"""
+    if retry=True, then reseed with result of easier MCS(RemoveHs(a), RemoveHs(b)) in case of failure
+    """
     params = rdFMCS.MCSParameters()
+
+    # bonds
     params.BondCompareParameters.CompleteRingsOnly = 1
     params.BondCompareParameters.RingMatchesRingOnly = 1
     params.BondTyper = rdFMCS.BondCompare.CompareAny
 
+    # atoms
     params.AtomCompareParameters.CompleteRingsOnly = 1
     params.AtomCompareParameters.RingMatchesRingOnly = 1
     params.AtomCompareParameters.matchValences = 0
-    params.AtomCompareParameters.MaxDistance = threshold
+    params.AtomCompareParameters.MatchChiralTag = 0
+    if conformer_aware:
+        params.AtomCompareParameters.MaxDistance = threshold
     params.AtomTyper = rdFMCS.AtomCompare.CompareAny
 
+    # globals
     params.Timeout = timeout
     if smarts is not None:
         params.InitialSeed = smarts
-    return rdFMCS.FindMCS([a, b], params)
-
-
-def mcs_map_graph_only_complete_rings(a, b, timeout: int = 3600, smarts: Optional[str] = None):
-    """Find the MCS map of going from A to B, disregarding conformer information. This also ensures
-    that core-core bonds are not broken."""
-    return rdFMCS.FindMCS(
-        [a, b],
-        maximizeBonds=True,
-        timeout=timeout,
-        matchValences=False,
-        ringMatchesRingOnly=True,
-        completeRingsOnly=True,
-        matchChiralTag=False,
-        atomCompare=Chem.rdFMCS.AtomCompare.CompareAny,
-        bondCompare=Chem.rdFMCS.BondCompare.CompareAny,
-    )
+
+    # try on given mols
+    result = rdFMCS.FindMCS([a, b], params)
+
+    # optional fallback
+    def is_trivial(mcs_result) -> bool:
+        return mcs_result.numBonds < 2
+
+    if retry and is_trivial(result) and smarts is None:
+        # try again, but seed with MCS computed without explicit hydrogens
+        a_without_hs = Chem.RemoveHs(deepcopy(a))
```

Comment:
Actually from reading the docs, this always returns a copy anyway (https://www.rdkit.org/docs/source/rdkit.Chem.rdmolops.html)

RETURNS: a new molecule with the Hs removed

NOTES:

The original molecule is not modified.

Reply:
Even so, I think the extra few microseconds it costs to `deepcopy` is worth it to avoid having a reader ever be prompted to confirm this, and I'd be inclined to keep the redundant `deepcopy` here.